New components: SidePanel & SimpleModal

tinkertrain opened the pull request:
# Feature

Design specs: https://www.figma.com/file/v8ReousbOpHN2g9n5n6StX/Standalone-Messages---2.0?node-id=342%3A2552

## SidePanel

We add a new overlay component: SidePanel. This is sort of a modal that slides from the side.

[Docs and stories](https://deploy-preview-962--hsds-react.netlify.app/?path=/docs/components-overlay-sidepanel--default-story)

### SidePanel Layouts

We have some layouts that are going to be commonly used, in order to avoid duplication we include what we call "layouts" which are simple components that you can drop in the `<SidePanel>` and forget about styling. They live inside `SidePanel.layouts.jsx`.

#### HeaderAndFooter layout

Includes:

- Header with a heading and a sub-heading
- Scrollable body (`children` get rendered here). Shadows get added to header and footer depending on the scroll state of the body.
- footer with a CTA.

Example usage:

```jsx
import { HeaderAndFooter } from './SidePanel.layouts'

function YourApp() {
  return (
    <SidePanel>
      <HeaderAndFooter
        mainActionButtonText="Customize this"
        panelHeading="Customize this"
        panelSubHeading="Customize this"
        onMainActionClick={handleClick}
        mainActionDisabled={!isDisabled}
      >
        <div>Your content here</div>
      </HeaderAndFooter>
    </SidePanel>
  )
}
```

## SimpleModal

This is a new Modal component. We already have a Modal component that is bloated and would have required more work to achieve what we needed.

The hope is that eventually we use this as a default Modal and phase out the existing Modal, which is a mess and everybody hates it. One idea is to bring a concept similar to the DropList "togglers" in which we provide default layouts for the most common modal types we use (warning, confirmation, etc) that way we can keep adding to that list without adding a gazillion different props (some of which don't work together) and keep the underlying component as lean as possible.

[Docs and stories](https://deploy-preview-962--hsds-react.netlify.app/?path=/docs/components-overlay-simplemodal--default-story)

## Considerations

Both components are basically the same in terms of implementation, here are some of the characteristics of their design:

- Both are fully controlled, they don't store any state about their opened/closed status. Since the user will normally have to create some state that holds this in their app, it seems unnecessary to add the complexity of keeping the `show` prop and an internal piece of state in sync.
	- Use the `onClose` to close the component, it gets fired when clicking the <kbd>x</kbd> and on hitting <kbd>esc</kbd>
- They both will be contained within the most immediate parent with a `position: relative` applied to them, this allows you to create layouts where the SidePanel does not cover the navigation or to open the a modal that only covers said SidePanel, just like the design spec requires.
- If you need different styles, use Styled Components, there's nothing preventing you to do this and keeps all the props to the minimum needed.
- If you need to portal the components, either use the existing Portal component or the newly added `usePortal` hook.

![Full example](https://i.hlp.sc/MaZoOh)

## Custom hooks

We add a new root-level section "hooks" where custom shareable hooks live.

### useAnimatedRender

A lot of the times we need to animate a component before it gets mounted and unmounted. On mounting is easy enough, but for unmounting you need to hold the unmount, animate and then continue with the unmounting.

This hook allow you to do that, both SidePanel & SimpleModal use this hook, see those for examples of application and the documentation provided for more info.

[Full Docs](http://localhost:8900/?path=/story/hooks-useanimatedrender--page)

### useClickOutside

Custom hook that adds a click event listener on the document that gets executed when clicking outside a given DOM Node passed as a ref.

[Full Docs](http://localhost:8900/?path=/story/hooks-useclickoutside--page)

### usePortal

We already have a Portal component, here we add a hook that you can bring into any component if you prefer using a hooks solution. The implementation is leaner and has a couple of advantages like creating a container with the selector provided if it doesn't exist.

It includes a Portal component for convenience, which you can import from the same file.

[Docs and stories](http://localhost:8900/?path=/story/hooks-useportal--page)

## Documentation

Everything is documented, and you'll see a new hooks section on storybook.

jakubjanczyk commented on src/components/SidePanel/SidePanel.stories.mdx:
```diff
@@ -0,0 +1,85 @@
+
+Component that brings a Panel from either side of the screen, it includes a default Header and Footer (both can be turned off via their respective props if you need to fully customize the panel's contents). The footer contains a CTA button.
+
+This component is fully controlled via the `show` prop, meaning it doesn't hold any internal state about whether is open or not, it's in your hands to manage this. The panel will fire the close action (`onClose` callback prop) with the &times; close button and via pressing the <kbd>esc</kbd> key.
```
I really like that! I mean, a lot of components in HSDS have some internal state as a duplication of provided props and in most of the cases it only adds complexity. Having it outside makes that this component is quite simple :)

jakubjanczyk commented on src/components/SidePanel/SidePanel.test.js:
```diff
@@ -0,0 +1,203 @@
+    expect(queryByTestId('sidepanel-overlay')).toBeInTheDocument()
+    expect(
+      queryByTestId('sidepanel-overlay').classList.contains('element-in')
+    ).toBe(true)
```
What do you think about using `toHaveClass` assertion function? It would give better error message if failed :)

tinkertrain replied:
I wonder if we have that in HSApp, it's a good idea, though I'd like to add those in a different PR and perhaps add them to HSApp (or bring them from there)

jakubjanczyk replied:
I have been using them, so they are for sure :smile: It's from `@testing-library/jest-dom` library that is set up in `config/jest/setupTests.js` in hsds repo already

tinkertrain replied:
Talk about not knowing what we have... thanks! (I changed the assertions)

jakubjanczyk commented on src/components/SidePanel/SidePanel.test.js:
```diff
@@ -0,0 +1,203 @@
+
+    expect(heading).toBeInTheDocument()
+    expect(heading.getAttribute('id')).toBe('sidepanel-header-heading')
+    expect(heading.textContent).toBe('Review and Start')
```
Similar to above, in those 2 lines we could use `toHaveAttribute` and `toHaveTextContent` as well. And the same could be applied to some tests below as well.

tinkertrain replied:
ditto

jakubjanczyk commented on src/hooks/useAnimatedRender.js:
```diff
@@ -0,0 +1,65 @@
+      if (!show) {
+        setRender(false)
+
+        if (toAnimateRef.current) {
```
I think it would never be false? if `e.target === toAnimateRef.current` is true, and I think `target` would never be undefined/null (this is the place I have not checked, but my intuition tells me that :smile:), then there should be no need to have this additional check

tinkertrain replied:
Good catch!

jakubjanczyk reviewed (commented):
Looks and works great! 
I have only found one visual issue with scrollbars appearing for a moment while animation is running, you can see it here: https://c.hlp.sc/P8uGYL8K. I've checked in Chrome and Firefox, both looks the same. It's not a big issue, but would be good if can be fixed to avoid the jump on the page.

Charca commented on src/components/SidePanel/SidePanel.jsx:
```diff
@@ -0,0 +1,153 @@
+        panelWidth={panelWidth}
+        ref={panelRef}
+        onFocus={e => {
+          console.log('FOCUSED')
```
Stray console.log here. Also, should `onFocus` come as a prop to this component?

tinkertrain replied:
This one is leftover from debugging... removed!

Charca commented on src/components/SidePanel/SidePanel.jsx:
```diff
@@ -0,0 +1,153 @@
+        className="SidePanel"
+        data-cy="sidepanel"
+        data-testid="sidepanel"
+        id="sidepanel"
```
Do we need an ID here? Looks like a hard-coded ID will restrict us to a single instance of the side panel on each page.

tinkertrain replied:
I added the `{...rest}` so you can override any of those attributes on the Panel element

Charca commented on src/components/SimpleModal/SimpleModal.jsx:
```diff
@@ -0,0 +1,101 @@
+        className="SimpleModal"
+        data-cy="simple-modal"
+        data-testid="simple-modal"
+        id="simple-modal"
```
Same question about the ID here.

tinkertrain replied:
Same as above :)

plbabin reviewed (approved):
I really like that we start creating custom hook to help improve other component within hs-app. I think the `usePortal` one could make his way into other hsds component as well. 

I did also had the scrollbar issue with the SidePanel story. I think we could have an optional prop to "fixed" the SidePanel on the screen (or we let dev implement that if needed by overwriting the css with styled-components). Either way I am fine with that. 


Both component looks really good, they are minimal in their implementation and I like that!

Charca reviewed (commented):
Looks great @tinkertrain, fantastic job! I left a couple of minor comments here and there.

tinkertrain commented on src/components/SidePanel/SidePanel.jsx:
```diff
@@ -0,0 +1,153 @@
+        panelWidth={panelWidth}
+        ref={panelRef}
+        onFocus={e => {
+          console.log('FOCUSED')
```
Yes... thanks!

tinkertrain commented:
@jakubjanczyk @plbabin @Charca 

I managed to get rid of the scrollbars I think, let me know!

Also, sorry to move the goal posts here... but I modified a bit the component, simplifying further.

I added the concept of "component layouts"... let me know what you think! (more info on the description and the documentation on storybook)

jakubjanczyk commented on src/components/SidePanel/SidePanel.layouts.jsx:
```diff
@@ -0,0 +1,125 @@
+  let isTopScrolled = false
+  let isBottomScrolled = false
+
+  if (scrollable.scrollTop - scrollablePaddingTop > 0) {
```
What do you think about using ternary operator here, so that there's no need to create `let` variables? I like it more because it's more concise, but not something that is blocking :)

tinkertrain replied:
Ahh actually variables are not even needed, updated :)

jakubjanczyk reviewed (approved):
Looks good! No more scrolls for me and I like the concept of easily replaceable content!
I've left one question, but it's good to go :)

plbabin commented on src/components/SidePanel/SidePanel.stories.mdx:
````diff
@@ -0,0 +1,117 @@
+}
+```
+
+## SidePanel Layouts
````
oooh I like that, 💯  sure layouts will make his way to Modal 😄
